File: neetcode/two.sum.mjs
import assert from "node:assert";
import test from "node:test";

/**
 * Input:  nums = [3,4,5,6], target = 7
 * Output: [0,1]
 */
function twoSum(nums, target) {
  const map = {};

  for (let i = 0; i < nums.length; i++) {
    const diff = target - nums[i];

    if (diff in map) {
      return [map[diff], i];
    } else {
      map[nums[i]] = i;
    }
  }

  return [-1, -1];
}

test("test case 1", () => {
  assert.deepEqual(twoSum([3, 4, 5, 6], 7), [0, 1]);
});

test("test case 2", () => {
  assert.deepEqual(twoSum([4, 5, 6], 10), [0, 2]);
});

File: golang/two_sum.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	m := make(map[int]int)

	for i := 0; i < len(nums); i++ {
		r := target - nums[i]

		val, ok := m[r]

		if ok {
			return []int{i, val}
		} else {
			m[nums[i]] = i
		}
	}

	return []int{0, 0}
}

func main() {
	// Пример использования
	nums := []int{2, 7, 11, 15}
	target := 9
	result := twoSum(nums, target)

	fmt.Printf("Input: nums = %v, target = %d\n", nums, target)
	fmt.Print("Output: ")
	PrintArray(result) // Используем функцию из utils.go

	// Другой пример использования функции из utils.go
	fmt.Printf("Max value in nums: %d\n", Max(nums))
}
